File: app/src/main/java/com/mavenkalabs/adskipper/rules/NoRecentUserClickRule.java
package com.mavenkalabs.adskipper.rules;

import android.view.accessibility.AccessibilityNodeInfo;

import androidx.annotation.NonNull;

import java.util.Map;

import static com.mavenkalabs.adskipper.rules.RuleConstants.RULE_ID_NO_RECENT_USER_CLICK;
import static com.mavenkalabs.adskipper.rules.RuleConstants.RULE_PARAM_LAST_USER_CLICK_TS;

public class NoRecentUserClickRule implements BaseRule {
    private final long interval;

    public NoRecentUserClickRule(String ruleAsString) {
        String[] arr = ruleAsString.split(",");
        if (arr.length != 2) {
            throw new IllegalArgumentException();
        } else {
            interval = Long.parseLong(arr[1]);
        }
    }

    @Override
    public RuleResult apply(AccessibilityNodeInfo node, Map<String, Object> parameters) {
        if (parameters != null) {
            Object value1 = parameters.get(RULE_PARAM_LAST_USER_CLICK_TS);
            if (value1 instanceof Long) {
                long clickTS = (Long) value1;
                if (System.currentTimeMillis() > (clickTS + interval)) {
                    return new RuleResult(true);
                }
            }
        }
        return new RuleResult(false);
    }

    @NonNull
    @Override
    public String toString() {
        return String.join(",", RULE_ID_NO_RECENT_USER_CLICK, String.valueOf(interval)) ;
    }
}

File: app/src/main/java/com/mavenkalabs/adskipper/rules/NoRecentUserClickRule.kt
package com.mavenkalabs.adskipper.rules

import android.view.accessibility.AccessibilityNodeInfo

class NoRecentUserClickRule(ruleAsString: String) : BaseRule {
    private val interval: Long

    init {
        val arr: Array<String?> =
            ruleAsString.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        require(arr.size == 2)
        interval = arr[1]!!.toLong()
    }

    override fun apply(
        node: AccessibilityNodeInfo,
        parameters: Map<String, Any>?
    ): RuleResult {
        if (parameters != null) {
            val value1 = parameters.get(RuleConstants.Companion.RULE_PARAM_LAST_USER_CLICK_TS)
            if (value1 is Long) {
                val clickTS = value1
                if (System.currentTimeMillis() > (clickTS + interval)) {
                    return RuleResult(true)
                }
            }
        }
        return RuleResult(false)
    }

    override fun toString(): String {
        return listOf(
            RuleConstants.Companion.RULE_ID_NO_RECENT_USER_CLICK,
            interval.toString()).joinToString(",")
    }
}
